src/HydraAuctionOnchain/MintingPolicies/Auction.hs

```diff
@@ -0,0 +1,126 @@
+
+auctionMintingPolicy
+  :: Term
+      s
+      ( PScriptHash
+          :--> PTxOutRef
+          :--> PAuctionMpRedeemer
+          :--> PScriptContext
+          :--> PUnit
+      )
+auctionMintingPolicy = phoistAcyclic $
+  plam $ \auctionMetadataSh nonceOref redeemer ctx -> P.do
+    -- Script purpose should contain own currency symbol. Always
+    -- true for minting policies.
+    auctionCs <-
+      plet $
+        passertMaybe
+          $(errCode AuctionMp'Error'MissingOwnCurrencySymbol)
+          (pownCurrencySymbol # ctx)
+
+    pmatch redeemer $ \case
+      MintAuctionRedeemer _ ->
+        pcheckMintAuction # auctionMetadataSh # nonceOref # ctx # auctionCs
+      BurnAuctionRedeemer _ ->
+        pcheckBurnAuction # ctx # auctionCs
+
+pcheckMintAuction
+  :: Term
+      s
+      ( PScriptHash
+          :--> PTxOutRef
+          :--> PScriptContext
+          :--> PCurrencySymbol
+          :--> PUnit
+      )
+pcheckMintAuction = phoistAcyclic $
+  plam $ \auctionMetadataSh nonceOref ctx auctionCs -> P.do
+    txInfo <- plet $ pfield @"txInfo" # ctx
+    txInfoFields <- pletFields @["mint", "inputs"] txInfo
+    mintValue <- plet $ pnormalize # txInfoFields.mint
+    let inputs = txInfoFields.inputs
+
+    -- The auction state, auction metadata, and standing bid tokens
+    -- of the auction should all be minted.
+    -- No other tokens should be minted or burned.
+    passert $(errCode AuctionMp'MintAuction'Error'AuctionTokenBundleNotMinted) $
+      mintValue #== pauctionTokenBundleMinted # auctionCs
+
+    -- The utxo nonce parameter of the minting policy should be a
+    -- reference to a utxo input spent by the transaction.
+    passert $(errCode AuctionMp'MintAuction'Error'MissingUtxoNonceInput) $
+      pisJust #$ pfindTxInByTxOutRef # nonceOref # inputs
+
+    -- There should be exactly one auction metadata output.
+    auctionMetadataOutput <-
+      plet $
+        passertMaybe
+          $(errCode AuctionMp'MintAuction'Error'MissingAuctionMetadataOutput)
+          (pfindUniqueOutputWithScriptHash # auctionMetadataSh # txInfo)
+
+    -- The auction metadata output should contain an auction
+    -- metadata token.
+    passert $(errCode AuctionMp'MintAuction'Error'AuctionMetadataOutputMissingToken) $
+      ptxOutContainsAuctionMetadataToken # auctionCs # auctionMetadataOutput
+
+    -- The auction metadata output contains a datum that can be
+    -- decoded as an auction info metadata record.
+    (auctionInfo :: Term s PAuctionInfo) <-
+      plet $
+        passertMaybe
+          $(errCode AuctionMp'MintAuction'Error'FailedToDecodeAuctionInfo)
+          (pdecodeInlineDatum # auctionMetadataOutput)
+
+    -- The auction info metadata record should contain an auction ID
+    -- that matches this minting policy's currency symbol.
+    auctionInfoFields <- pletFields @["auctionId", "auctionTerms"] auctionInfo
+    passert $(errCode AuctionMp'MintAuction'Error'AuctionInfoCurrencySymbolMismatch) $
```

Comment:
I think this check wasn't originally there, shall we update the docs in hydra-auction? Well, the spec is missing not only this check. I think we need to update, I can do it, what do you think?

Reply:
Yes, there are multiple checks that we added in the course of development.